subsys/bluetooth/host/audio/ascs.c

```diff
@@ -56,6 +56,251 @@ struct bt_ascs {
+	qos->rtn = ep->stream->qos->rtn;
+	qos->latency = sys_cpu_to_le16(ep->stream->qos->latency);
+	sys_put_le24(ep->stream->qos->pd, qos->pd);
+
+	BT_DBG("dir 0x%02x codec 0x%02x interval %u framing 0x%02x phy 0x%02x "
+	       "rtn %u latency %u pd %u",
+	       ASE_DIR(ep->status.id), ep->stream->codec->id,
+	       ep->stream->qos->interval, ep->stream->qos->framing,
+	       ep->stream->qos->phy, ep->stream->qos->rtn,
+	       ep->stream->qos->latency, ep->stream->qos->pd);
+}
+
+static void ascs_ep_get_status_enable(struct bt_audio_ep *ep,
+				      struct net_buf_simple *buf)
+{
+	struct bt_ascs_ase_status_enable *enable;
+
+	enable = net_buf_simple_add(buf, sizeof(*enable));
+	enable->cig_id = ep->cig_id;
+	enable->cis_id = ep->cis_id;
+
+	enable->metadata_len = buf->len;
+	ascs_codec_data_add(buf, "meta", ep->codec.meta_count, ep->codec.meta);
+	enable->metadata_len = buf->len - enable->metadata_len;
+
+	BT_DBG("dir 0x%02x cig 0x%02x cis 0x%02x",
+	       ASE_DIR(ep->status.id), ep->cig_id, ep->cis_id);
+}
+
+static int ascs_ep_get_status(struct bt_audio_ep *ep,
+			      struct net_buf_simple *buf)
+{
+	struct bt_ascs_ase_status *status;
+
+	if (!ep || !buf) {
+		return -EINVAL;
+	}
+
+	BT_DBG("ep %p id 0x%02x state %s", ep, ep->status.id,
+		bt_audio_ep_state_str(ep->status.state));
+
+	/* Reset if buffer before using */
+	net_buf_simple_reset(buf);
+
+	status = net_buf_simple_add_mem(buf, &ep->status,
+					sizeof(ep->status));
+
+	switch (ep->status.state) {
+	case BT_AUDIO_EP_STATE_IDLE:
+	/* Fallthrough */
+	case BT_AUDIO_EP_STATE_RELEASING:
+		break;
+	case BT_AUDIO_EP_STATE_CODEC_CONFIGURED:
+		ascs_ep_get_status_config(ep, buf);
+		break;
+	case BT_AUDIO_EP_STATE_QOS_CONFIGURED:
+		ascs_ep_get_status_qos(ep, buf);
+		break;
+	case BT_AUDIO_EP_STATE_ENABLING:
+	/* Fallthrough */
+	case BT_AUDIO_EP_STATE_STREAMING:
+	/* Fallthrough */
+	case BT_AUDIO_EP_STATE_DISABLING:
+		ascs_ep_get_status_enable(ep, buf);
+		break;
+	default:
+		BT_ERR("Invalid Endpoint state");
+		break;
+	}
+
+	return 0;
+}
+
+static void ascs_iso_recv(struct bt_iso_chan *chan,
+			  const struct bt_iso_recv_info *info,
+			  struct net_buf *buf)
+{
+	struct bt_audio_ep *ep = CONTAINER_OF(chan, struct bt_audio_ep, iso);
+	struct bt_audio_stream_ops *ops = ep->stream->ops;
+
+	BT_DBG("stream %p ep %p len %zu", chan, ep, net_buf_frags_len(buf));
+
+	if (ops != NULL && ops->recv != NULL) {
+		ops->recv(ep->stream, buf);
+	}
+}
+
+static void ascs_iso_connected(struct bt_iso_chan *chan)
+{
+	struct bt_audio_ep *ep = CONTAINER_OF(chan, struct bt_audio_ep, iso);
+	struct bt_audio_stream_ops *ops = ep->stream->ops;
+
+	BT_DBG("stream %p ep %p type %u", chan, ep, ep != NULL ? ep->type : 0);
+
+	if (ops != NULL && ops->connected != NULL) {
+		ops->connected(ep->stream);
+	}
+
+	if (ep->status.state != BT_AUDIO_EP_STATE_ENABLING) {
+		BT_DBG("endpoint not in enabling state: %s",
+		       bt_audio_ep_state_str(ep->status.state));
+		return;
+	}
+
```

Comment:
Should the connected callback rather be called _after_ the check for state?
As it now is, the callback will be called also if the state is not valid.
And the callback may be called even if the state is not set to STREAMING

Reply:
The `connected` callback notifies about an ISO connection, regardless of the ASE state, so this is correct. 

> As it now is, the callback will be called also if the state is not valid.

It isn't invalid; the client may connect the ISO at any point.